Feature/#6 QnA 조회 API 구현

jher235 opened the pull request:
# 💥 Work Description
- 무엇에 관한 PR 인가요?
- 어떤 작업을 진행했나요?
- QnA 조회 API 구현
  - [x] QnA 도메인 추가
  - [x] QnA JPA Repository 인터페이스 구현
  - [x] QnA 목록 조회 구현
  - [x] QnA 상세 조회 구현 

# ⚙️ ISSUE
- closed #6


# 💡 변경사항
- QnA 엔티티, Image 엔티티 작성
- QnA 관련 컨트롤러, 서비스와 ResponseDto 를 작성


# 🌱 PR Point
- 리뷰어들에게 하고 싶은 말이나 중점적으로 봐주었으면 하는 것


# 🔗 Reference
문제를 해결하면서 도움이 되었거나, 참고했던 사이트(코드링크)

jher235 committed:
Feature:  QnA 엔티티 및 Image 엔티티 작성 - #6

jher235 committed:
Feature: QnARepository 작성 - #6
- Department로 QnA를 찿는 메서드 작성

jher235 committed:
Feature: QnA 조회 시 사용하는 ResponseDto 작성 - #6
- QnAPreviewResponse 는 빌더 패턴 생성자 사용
- 정적 팩토리 메서드 from 사용

jher235 committed:
Feature: LocalDateTime을 TimeElapsed로 변환하는 메서드 작성 - #6

jher235 committed:
Feature: QnA 목록 조회 구현 - #6
- 쿼리 스트링에는 department, size, page를 사용 가능하며 모두 필수사항은 아님.

jher235 committed:
Feature: Department enum 작성 - #6

jher235 committed:
Merge branch 'main' into feature/#6

jher235 committed:
Refactor: remote main 을 받아서 merge 하면서 main의 Department 를 사용하도록 수정 - #6

jher235 committed:
Feature: department를 한글로 반환하도록 수정 - #6

jher235 committed:
Feature: QnA 관련 예외 클래스 및 예외 코드 작성 - #6

jher235 committed:
Feature: QnA 상세 조회 목록을 반환할 QnADetailResponse 작성 - #6

jher235 committed:
Feature: QnA 상세 조회 목록 로직 작성 - #6

jher235 committed:
Feature: QnA 전체 목록 조회를 위한 Dto 작성 - #6
- 페이지네이션을 사용하지 않음

jher235 committed:
Feature: 최신순으로 모든 QnA를 찾는 메서드 작성 - #6
- PK 생성 전략이 단조 증가이므로 createdAt을 대신하여 클래스터링 인덱스를 사용함.

jher235 committed:
Feature: QnA 전체 조회 구현 - #6
- 사전에 프론트와 이야기된 형식은 해당 로직과 같이 페이지네이션을 사용하지 않는 형태였으므로 이런 api를 구현함.
- 다만 페이지네이션, 필터링 등을 클라이언트 단에서 진행하는 것이 효율적이지 않다고 판단하여 페이지네이션, 필터링을 진행하는 api도 구현한 상태임.

ssangbaam commented:
분량이 많으셨을텐데,,고생 많으셨습니다!

ssangbaam commented on src/main/java/sopt/sopt36goodoc/qna/service/QnAService.java:
```diff
@@ -0,0 +1,51 @@
+
+        if(departmentString != null && !departmentString.isBlank()){
+            Department department = Department.findByEnglishName(departmentString);
+            Page<QnA> qnAS = qnARepository.findAllByDepartment(pageable, department);
```
페이지를 받아온 후에, 존재하지 않는 페이지인 경우 빈 페이지로 뜨긴 하는데 좀 더 명확하게 페이지에 관한 예외처리가 필요해 보입니다!

jher235 replied:
빈 페이지를 오류로 처리하는게 좋다고 생각하신 이유가 있을까요?? 
기획에 따라 다르겠지만 개인적으로는 클라이언트 입장에선 데이터가 없는 경우 빈 리스트로 반환받는 게 처리하고 편하기도 하고 비어있는 페이지를 받는 것 자체에는 큰 문제가 없다고 생각하는데요. (물론 음수가 오지 못하도록 체크하는 부분은 필요하다고 생각합니다) ㅎㅎ 상범님 생각은 어떠신가요??
++ 우선 예외처리는 진행하겠습니다 ~

ssangbaam replied:
제가 생각하기에는, 잘못된 페이지 요청인 경우, 예를 들어, 전체가 3페이지인데 100페이지 요청을 보낸 경우에 정상 응답(200OK)이 나온다면 유효한 요청처럼 보일 수도 있어 예외처리가 필요하다 생각했습니다!

ssangbaam commented on src/main/java/sopt/sopt36goodoc/qna/entity/QnA.java:
혹시 여기 패키지명을 entity로 한 다른 이유가 있을까요?

jher235 replied:
사실 로직 자체가 존재하지 않고 단순 매핑을 하고 있기 때문에 entity로 두었는데 통일성을 위해서 domain으로 수정하겠습니다~

ssangbaam commented on src/main/java/sopt/sopt36goodoc/qna/dto/response/AllQnAPreviewResponse.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+
+public record AllQnAPreviewResponse(List<QnAPreviewResponse> qnAPreviews) {
+    public static AllQnAPreviewResponse from(List<QnA> qnAS){
```
qnAS에서 S가 의도한게 대문자가 맞는지 궁금합니다!

jher235 replied:
앗 오타입니다..! 바로 수정할게요 ㅎㅎ

jher235 committed:
Refactor: entity 패키지명을 domain으로 수정 - #6
- 통일성을 위함

jher235 committed:
Style: 카멜케이스 오타 수정 - #6

jher235 committed:
Feature: HandlerMethodValidationException 핸들링 메서드 추가 - #6
- Spring 3.2 부터 MethodValidator가 Valid를 처리하므로 해당 부분이 필요함. 특히 현재는 @RequestParam의 Valid를 관리하기 위함

jher235 committed:
Feature: 메세지를 찾지 못한 경우의 예외코드 추가 - #6

jher235 committed:
Feature: 최소 사이즈가 0이 되도록 Valid 추가 - #6

jher235 committed:
Feature: 존재하지 않는 페이지 요청 시의 검증 추가 - #6
- checkValidPageRequest 적용 부분을 함수형 인터페이스를 적용하여 수행하도록 함

ssangbaam commented on src/main/java/sopt/sopt36goodoc/global/exception/GlobalErrorCode.java:
```diff
@@ -24,6 +24,8 @@ public enum GlobalErrorCode implements ErrorCode{
     NOT_SUPPORTED_METHOD_ERROR(HttpStatus.METHOD_NOT_ALLOWED, "REQUEST_004", "지원하지 않는 HTTP Method 요청입니다."),
     NOT_SUPPORTED_MEDIA_TYPE_ERROR(HttpStatus.UNSUPPORTED_MEDIA_TYPE, "REQUEST_005", "잘못된 미디어 타입입니다."),
+
+    NOT_FOUND_PAGE(HttpStatus.NOT_FOUND, "PAGE_001", "페이지를 찾을 수 없습니다."),
```
GlobalErrorCode에 포함시키기보다 global 패키지 안에 별도의 PageErrorCode enum으로 분리해 관리하는 것이 구조적으로 더 괜찮을것 같안데 어떻게 생각하시나요??!

jher235 replied:
Page는 Global로 바라보는게 더 좋지 않을까요?? 만약 PageErrorCode를 나눈다면 이걸 나누는 기준은 어떻게 될까요?? 현재는 REQUEST 관련 오류도, RESPONSE 관련 오류도 GlobalErrorCode에서 처리하고 있는데 여기서는 유효성 검사, 인가, 서버 인스턴스 문제 등의 오류도 함께 다루고 있기 때문에 Page 부분만 따로 떼는게 필요할까? 라는 생각입니다. 
아니면 차라리 PAGE_001을 REQUEST_006 정도로 두는 것은 어떨까요??

ssangbaam replied:
제가 고민한 이유는 Page 자체는 도메인 객체는 아니지만, 명확한 에러 범주로 구분할 수 있어 별도의 PageErrorCode로 관리하면 좋겠다 생각했습니다!
말씀해주신 대로 page관련 에러가 전역적으로 사용되기 때문에 GlobalErrorCode에 포함시켜도 될 것 같습니다!

jher235 committed:
Feature: 에러 코드 수정 - #6
- PAGE_001 -> REQUEST_006 으로 수정
